LPC55S69: TFM with MCUboot to run Zephyr NS rework

agansari opened the pull request:
Enables LPCXpresso55S69 board to run TFM with MCUboot.
 - fit dts's flash partitioning to TFM's flash_layout.h
 - remove BL2=OFF as default from board
 - merge all binaries to a single image: tfm_merged.bin
 - update all documentation to reflect new binary
 - LPC55S69 requires TFM_PROFILE = profile_medium to fit it's flash 

Depends on: #30061

agansari committed:
boards: lpcxpresso55s69: partition refactor

Refactor lpcxpresso55s69's partitions to match TFM's flash_layout.h
configuration. This matches TFM with MCUboot configuration.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: lpcxpresso55s69: BL2=OFF memory settings

Create a special configuration when BL2=OFF is set. DTS partitioning is
used for MCUboot, but does not match TFM's flash_layout.h configuration
when BL2=OFF, DTS matches when BL2=ON.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
boards: lpcxpresso55s69: enable TFM with MCUboot

Set building TFM with MCUboot. Set the build configuration to
profile_medium, we need smaller TFM images to fit into flash.
Build MCUboot, TFM, sign it, sign Zephyr NS image and merge all the
images. Also change the other configuration, BL2=OFF, to merge as a
single image.
Update documentation on how to flash the board.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari committed:
samples: tfm_integration: lpc55s69 documentation

Update sample documentations to reflect the new way to flash
lpcxpresso55s69 boards with TFM.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

microbuilder commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -33,4 +36,65 @@ if (CONFIG_BUILD_WITH_TFM)
+			#Copy mcuboot.bin
+			COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_BINARY_DIR}/tfm/bin/bl2.bin ${CMAKE_BINARY_DIR}/mcuboot.bin
+
+			#Merge mcuboot.bin and tfm_sign.bin for QEMU
```
Minor detail, but this chip doesn't support QEMU so just drop the `for QEMU` comment.

microbuilder commented on boards/arm/lpcxpresso55s69/doc/index.rst:
```diff
@@ -202,16 +202,20 @@ see the following message in the terminal:
+   .. code-block:: console
+
+      JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
+      J-Link>loadfile build/tfm_merged.bin
```
Should we add a note on the `r` hack in cases where you can't connect to the debug interface because TF-M has previously started, just to have it on record or explain the common error message?

MaureenHelm replied:
Why are there separate instructions for JLinkExe and west? If it's because of the erase command, you can use `west flash --erase`.

agansari replied:
@MaureenHelm the west flash command instruction was erased because it's not working as expected.

microbuilder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -134,16 +134,19 @@ Build Zephyr with a non-secure configuration:
       JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
-      J-Link>loadfile build/tfm/install/outputs/NXP/LPCXPRESSO55S69/tfm_s.hex
-      J-Link>loadfile build/zephyr/zephyr.hex
+      J-Link>loadfile build/tfm_merged.bin
```
Same issue as earlier, perhaps mention the need to reset in some situations before we can flash.

microbuilder commented:
@tejlmand @ioannisg Any opinion(s) on the proposed changes in `modules/Kconfig.tfm`?

ioannisg commented:
@agansari may i ask you to split out the parts of this PR that are not directly related to the LPC board (and the support for BL2)?
I.e. the parts that add and parse Kconfig variables into cmake build options. That will accelerate the reviewing of the patch.

ioannisg commented:
> @agansari may i ask you to split out the parts of this PR that are not directly related to the LPC board (and the support for BL2)?
> I.e. the parts that add and parse Kconfig variables into cmake build options. That will accelerate the reviewing of the patch.

I mean, split out those parts into a new PR :)

agansari commented:
@ioannisg #30061 was created for TFM, trimming this PR to be LPC only related.

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -33,4 +36,65 @@ if (CONFIG_BUILD_WITH_TFM)
+				--pad-header
+				${ADD_NS_IMAGE_MIN_VER}
+				-s auto
+				-H 0x400
```
Not sure, but would this value happen to be this setting ?:
https://github.com/zephyrproject-rtos/zephyr/blob/35264cc214fd37e6da8cc682b4f47e86342f517a/modules/Kconfig.tfm#L116-L118

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -33,4 +36,65 @@ if (CONFIG_BUILD_WITH_TFM)
+				-v ${TFM_IMAGE_VERSION_NS}
+				-s auto
+				${ADD_S_IMAGE_MIN_VER}
+				-H 0x400
```
Not sure, but would this value happen to be this setting ?:
https://github.com/zephyrproject-rtos/zephyr/blob/35264cc214fd37e6da8cc682b4f47e86342f517a/modules/Kconfig.tfm#L116-L118

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -33,4 +36,65 @@ if (CONFIG_BUILD_WITH_TFM)
+			COMMAND ${SREC_CAT}
+			ARGS ${CMAKE_BINARY_DIR}/tfm/bin/tfm_s.bin -Binary
+				${CMAKE_BINARY_DIR}/zephyr/${KERNEL_BIN_NAME} -Binary
+				-offset 0x40000
```
Should this not use the Kconfig setting `${CONFIG_FLASH_LOAD_OFFSET}`? 
https://github.com/zephyrproject-rtos/zephyr/blob/f70e29281cb7c396891027180aefef3e0b5a6db1/boards/arm/lpcxpresso55s69/Kconfig.defconfig#L59-L60

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -33,4 +36,65 @@ if (CONFIG_BUILD_WITH_TFM)
+			COMMAND ${SREC_CAT}
+			ARGS ${CMAKE_BINARY_DIR}/mcuboot.bin -Binary
+				${CMAKE_BINARY_DIR}/tfm_sign.bin -Binary
+				-offset 0x8000
```
What is this value ?
If we don't have a config setting for this, then at least make a comment on why the value is 0x8000 and what it needs to match.
That would help the next guy reading this code.

tejlmand reviewed (commented):
Minor comments.

tejlmand commented on boards/arm/lpcxpresso55s69/CMakeLists.txt:
```diff
@@ -33,4 +36,65 @@ if (CONFIG_BUILD_WITH_TFM)
+				-offset 0x40000
+				-o ${CMAKE_BINARY_DIR}/tfm_merged.bin -Binary
+		)
+	else()
```
This will be executed if `CONFIG_TFM_BL2_TRUE=y` but also if `CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y` and the TFM config file has `TFM_BL2=False`.

I expect things will break in that case.

Should this have been an `elseif()` ?
And how is the `CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y` supposed to be handled ?

agansari replied:
@tejlmand i've updated CMakeLists.txt with kconfig values and added a comment where it was not possible.

> This will be executed if CONFIG_TFM_BL2_TRUE=y but also if CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y and the TFM config file has TFM_BL2=False.
 
This is not the case, `TFM_BL2` is set according to Zephyr's KConfig.
`CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y` is the default and how it was tested most of the time.
`CONFIG_TFM_BL2_TRUE=y` was tested and has the same result.

`CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y` does not set `-DBL2` in TFM which meas, we use board's default in TFM. (All boards have `BL2=ON` as default).
`CONFIG_TFM_BL2_TRUE/FALSE` forces the build with `-DBL2=ON/OFF`.

tejlmand replied:
Thanks for the explanation.

> CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y does not set -DBL2 in TFM which meas, we use board's default in TFM. (All boards have BL2=ON as default).

Maybe worth mentioning this in a comment ?

tejlmand reviewed (changes requested):
Please check behavior reqarding: `CONFIG_TFM_BL2_CONFIG_FILE_DEFAULT=y`

agansari commented:
We have a problem with building this image for the samples @microbuilder @ioannisg, so I added **DNM** label.
The problem lies with building 'tfm_ns.bin' image (which is not used) in the sample.

As I observe `-DTEST_NS` is the one making the `tfm_ns.bin` image larger than the NS flash partition.
Do we have a reason why this flag is set? It's changing an image we don't even use.

Do you propose any solution, I can think of:
 - set `-DTEST_NS=OFF` for this particular board or when building with Zephyr
 - make zephyr+tfm build without tfm_ns.bin image

Building the upstream TFM images with `-DTEST_NS=OFF` makes the image just as large. Can you help me understand what this macro does? LPC55S69 is memory constrained.

tejlmand reviewed (approved):
Approved, but consider to add some comments on the `else()` section regarding TFM BL2 default implementation.

agansari committed:
modules: tfm: BL2 default comment

Adds a comment regarding the effect of TFM_BL2_CONFIG_FILE_DEFAULT on
building TFM images.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

agansari commented:
@tejlmand added this description in Kconfig.tfm as it's something general.

agansari commented:
@microbuilder @ioannisg i've created pull https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/31
This commit removes the test code that enlarges the unused image **tfm_ns.bin** and shortens build time.

tejlmand commented on modules/Kconfig.tfm:
```diff
@@ -84,17 +84,20 @@ config TFM_ISOLATION_LEVEL
-        bool "Use TFM BL2 setting from TFM configuration file"
+	bool "Use TFM BL2 setting from TFM configuration file"
+	help
+	  This option does not set -DBL2 in TFM, which means board's default
+	  is used. Currently all boards have BL2=ON as default.
```
TF-M uses the term `platforms`, so maybe it is better to use TF-M Platform instead of board ?
https://ci-builds.trustedfirmware.org/static-files/MBqpg7FOB3e4jpcJW8Nfq0G9cXO3vR0hJC0xDODqZnAxNjA2MzAwMzI4NTQ4Ojk6YW5vbnltb3VzOnZpZXcvVEYtTS9qb2IvdGYtbS1idWlsZC1kb2NzLW5pZ2h0bHkvbGFzdFN0YWJsZUJ1aWxkL2FydGlmYWN0/trusted-firmware-m/build/install/doc/user_guide/html/platform/ext/index.html#platforms

Also helps the reader to distinguish between a Zephyr board, and a TF-M platform.

```suggestion
	  This option does not set -DBL2 in TFM, which means the TF-M platform's default
	  is used. Currently all TF-M platforms have BL2=ON as default.
```

tejlmand reviewed (commented):
Maybe better to use `TF-M platform` instead of `board`.

Relates to: https://github.com/zephyrproject-rtos/zephyr/pull/30145#issuecomment-730962083

erwango commented:
> @microbuilder @ioannisg i've created pull zephyrproject-rtos/trusted-firmware-m#31
This commit removes the test code that enlarges the unused image tfm_ns.bin and shortens build time.

@agansari isn't it required in order to run `tfm_ipc` sample ?

agansari commented:
@microbuilder @ioannisg please revisit this PR that enables MCUboot with TFM on LPCXpresso55S69.
It works now that TEST_NS is disabled by default.

microbuilder commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -140,16 +140,25 @@ Build Zephyr with a non-secure configuration:
 
-Next we need to manually flash the secure (``tfm_s.hex``)
-and non-secure (``zephyr.hex``) images wth a J-Link as follows:
+Make sure your board is set up with :ref:`lpclink2-jlink-onboard-debug-probe`.
+Next we need to manually flash the resulting image (``tfm_merged.bin``) with a
+J-Link as follows:
```
```suggestion
Make sure your board is set up with :ref:`lpclink2-jlink-onboard-debug-probe`,
since this isn't the debug interface boards ship with from the factory;

Next we either manually flash the resulting image (``tfm_merged.bin``) with a
J-Link as follows:
```

microbuilder commented on samples/tfm_integration/tfm_ipc/README.rst:
```diff
@@ -140,16 +140,25 @@ Build Zephyr with a non-secure configuration:
+Resetting the board and erasing it will unlock the board, this is useful in case
+it's in an unknown state and can't be flashed.
+
+or
```
```suggestion
... or we can use `west` to flash via:
```

microbuilder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -134,16 +134,25 @@ Build Zephyr with a non-secure configuration:
 
-Next we need to manually flash the secure (``tfm_s.hex``)
-and non-secure (``zephyr.hex``) images wth a J-Link as follows:
+Make sure your board is set up with :ref:`lpclink2-jlink-onboard-debug-probe`.
+Next we need to manually flash the resulting image (``tfm_merged.bin``) with a
+J-Link as follows:
```
```suggestion
Make sure your board is set up with :ref:`lpclink2-jlink-onboard-debug-probe`,
since this isn't the debug interface boards ship with from the factory:

Next we either manually flash the resulting image (``tfm_merged.bin``) with a
J-Link as follows:
```

microbuilder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -134,16 +134,25 @@ Build Zephyr with a non-secure configuration:
+Resetting the board and erasing it will unlock the board, this is useful in case
+it's in an unknown state and can't be flashed.
+
+or
```
```suggestion
... or we can use `west` to flash via:
```

microbuilder commented on boards/arm/lpcxpresso55s69/doc/index.rst:
```diff
@@ -202,16 +202,22 @@ see the following message in the terminal:
+   .. code-block:: console
+
+      JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
+      J-Link>reset
```
```suggestion
      J-Link>r
```

microbuilder commented on samples/tfm_integration/psa_level_1/README.rst:
```diff
@@ -134,16 +134,25 @@ Build Zephyr with a non-secure configuration:
       JLinkExe -device lpc55s69 -if swd -speed 2000 -autoconnect 1
-      J-Link>loadfile build/tfm/install/outputs/NXP/LPCXPRESSO55S69/tfm_s.hex
-      J-Link>loadfile build/zephyr/zephyr.hex
+      J-Link>reset
```
```suggestion
      J-Link>r
```

ioannisg commented on modules/Kconfig.tfm:
```diff
@@ -76,17 +76,20 @@ config TFM_ISOLATION_LEVEL
 choice TFM_BL2
-        prompt "BL2 configuration, should TFM build with MCUboot support"
-        default TFM_BL2_CONFIG_FILE_DEFAULT
+	prompt "BL2 configuration, should TFM build with MCUboot support"
```
Could you revert the white space diffs here, @agansari ?
Or are they actually "fixes"?

ioannisg replied:
multiple occasions in the file

agansari replied:
@ioannisg yes, it's a fix, there were different indentations in the file.
Do you see a problem with the new indentation?

MaureenHelm commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69.dtsi:
```diff
@@ -100,32 +100,32 @@
+			reg = <0x00070000 0x00018000>;
 		};
 		/*
 		 * The flash starting at 0x90000 and ending at
 		 * 0x9D7FF is reserved for the application.
```
This comment is now longer correct

agansari replied:
@MaureenHelm fixed the comment

ioannisg commented:
Taking a final look now..

ioannisg commented:
Good work @agansari , now all platforms in Zephyr that support TFM, have the same capabilities :)

Future work. we need to look into how to use DTree more for flash, ram partitioning for TFM, and clean up all these redefinitions of FLASH_LOAD_OFFSET and _SIZE.
